### PR TITLE
Unbreak mediaSource+facingMode on FF37 (regression) + constraints tests

### PR DESCRIPTION
**Pull request by jan-ivar**
#43 + requested tests.

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -22,6 +23,28 @@ test('log suppression', function(t) {
   t.end();
 });
 
+// Helpers to test adapter's legacy constraints-manipulation
```
Comment (KaptenJansson) [inline test/test.js]: Add punctuation.

File: test/test.js [hunk 2]
```diff
@@ -105,6 +130,186 @@ test('call getUserMedia with constraints', function(t) {
...
+      });
+      t.deepEqual(constraints, beforeAfter[1],
+                  'Constraints ' + (i + 1) + ' back-converted to ' + gum);
+    });
+  }
+
+  if (m.webrtcDetectedBrowser === 'firefox') {
+    pretendVersion(m, 37, function() {
+      testBeforeAfterPairs('mozGetUserMedia', [
+        // test that spec constraints get back-converted on FF37
```
Comment (KaptenJansson) [inline test/test.js]: Add punctuation.

File: test/test.js [hunk 3]
```diff
@@ -105,6 +130,186 @@ test('call getUserMedia with constraints', function(t) {
...
+             advanced:[
+               {width: {min: 1280, max: 1280}},
+               {height: {min: 720, max: 720}},
+               {facingMode: 'user'}
+             ],
+             require:['height', 'frameRate']
+           }
+         }
+        ],
+        // test that legacy constraints pass through unharmed on FF37
```
Comment (KaptenJansson) [inline test/test.js]: Add punctuation.

File: test/test.js [hunk 4]
```diff
@@ -105,6 +130,186 @@ test('call getUserMedia with constraints', function(t) {
...
+             ],
+             require:['height', 'frameRate']
+           }
+         }
+        ],
+      ]);
+    });
+    pretendVersion(m, 38, function() {
+      testBeforeAfterPairs('mozGetUserMedia', [
+        // test that spec constraints pass through unharmed on FF38+
```
Comment (KaptenJansson) [inline test/test.js]: Add punctuation.

File: test/test.js [hunk 5]
```diff
@@ -105,6 +130,186 @@ test('call getUserMedia with constraints', function(t) {
...
+             facingMode: 'user',
+             frameRate: {exact: 50}
+           }
+         },
+        ],
+      ]);
+    });
+  } else if (m.webrtcDetectedBrowser === 'chrome') {
+    testBeforeAfterPairs('webkitGetUserMedia', [
+      // test that spec constraints get back-converted on Chrome
```
Comment (KaptenJansson) [inline test/test.js]: Add punctuation.

File: test/test.js [hunk 6]
```diff
@@ -105,6 +130,186 @@ test('call getUserMedia with constraints', function(t) {
...
+           optional: [
+             {minWidth: 1280},
+             {maxWidth: 1280},
+             {minHeight: 720},
+             {maxHeight: 720},
+           ]
+         }
+       }
+      ],
+      // test that legacy constraints pass through unharmed on Chrome
```
Comment (KaptenJansson) [inline test/test.js]: Add punctuation.

File: test/test.js [hunk 7]
```diff
@@ -105,6 +130,186 @@ test('call getUserMedia with constraints', function(t) {
...
+           optional: [
+             {minWidth: 1280},
+             {maxWidth: 1280},
+             {minHeight: 720},
+             {maxHeight: 720},
+           ]
+         }
+       }
+      ],
+      // test code protecting Chrome from choking on common unknown constraints
```
Comment (KaptenJansson) [inline test/test.js]: Add punctuation.

File: test/test.js (hunk 1 above)
Comment (KaptenJansson) [inline test/test.js]: Add punctuation.

File: test/test.js [hunk 8]
```diff
@@ -22,6 +23,28 @@ test('log suppression', function(t) {
...
+// Helpers to test adapter's legacy constraints-manipulation
+
+function pretendVersion(m, version, func) {
+  var realVersion = m.webrtcDetectedVersion;
+  m.webrtcTesting.version = version;
+  func();
+  m.webrtcTesting.version = realVersion;
+}
+
+function mugGum(gum, func) {
```
Comment (KaptenJansson) [inline test/test.js]: Could you name this to something more descriptive?

**Comment by KaptenJansson** [comment]
I will let @alvestrand take a good look at the logic here, he should be back soonish.

**Comment by alvestrand** [comment]
Looks good to me. Merging.

**Pull request merged**